fedimint/src/lib.rs

```diff
@@ -250,26 +284,52 @@ impl FedimintServer {
         // last signed epoch is at most 3 epochs before the next epoch + faulty nodes because
         // faulty nodes can withhold sigs for an epoch before getting banned
         let max_age: u64 = self.cfg.max_faulty() as u64 + 3;
+        let threshold = self.cfg.max_faulty() * 2 + 1;
+
+        // include our expected next_epoch as well in case we can contribute to the next consensus
+        let last_saved = self.consensus.db.get_value(&LastEpochKey);
+        let next_epoch = last_saved.expect("DB error").map(|e| e.0 + 1).unwrap_or(0);
+        consensus_peers.insert(self.cfg.identity, next_epoch);
 
         loop {
-            // a threshold of peers sent verified epochs, so target the next epoch
-            if consensus_peers.len() > self.cfg.max_faulty() * 2 {
+            // if a threshold of peers agree on an epoch, go with that
+            for epoch in consensus_peers.values() {
+                if consensus_peers.values().filter(|e| *e == epoch).count() >= threshold {
+                    return *epoch;
+                }
+            }
```

Comment:
This feels like there should be a nicer abstraction for it. Maybe

```rust
    if let Some(epoch) = consensus_peers
        .values()
        .find(|epoch| consensus_peers.values().filter(|e| *e == epoch).count() >= threshold)
    {
        return epoch;
    }
```

But that also looks weird. Maybe this is the best way …

Reply:
I think my way was more readable but that's probably because I wrote it :)